indico/modules/events/registration/fields/sessions.py

```diff
@@ -0,0 +1,93 @@
+from indico.core.marshmallow import mm
+from indico.modules.events.registration.fields.base import RegistrationFormFieldBase
+from indico.modules.events.registration.models.registrations import RegistrationData
+from indico.modules.events.sessions.models.blocks import SessionBlock
+from indico.util.i18n import _
+
+
+class SessionsFieldDataSchema(mm.Schema):
+    minimum = fields.Integer(load_default=0, validate=validate.Range(0))
+    maximum = fields.Integer(load_default=0, validate=validate.Range(0))
+    collapse_days = fields.Bool(load_default=False)
+
+    @validates_schema(skip_on_field_errors=True)
+    def validate_min_max(self, data, **kwargs):
+        if data['minimum'] and data['maximum'] and data['minimum'] > data['maximum']:
+            raise ValidationError('Maximum value must be less than minimum value.', 'maximum')
+
+
+class SessionsField(RegistrationFormFieldBase):
+    name = 'sessions'
+    mm_field_class = fields.List
+    mm_field_args = (fields.Integer,)
+    setup_schema_base_cls = SessionsFieldDataSchema
+
+    @property
+    def default_value(self):
+        return []
+
+    @property
+    def filter_choices(self):
+        sessions_ids = [s.id for s in self.form_item.registration_form.event.sessions]
+        return {str(s.id): s.full_title for s in SessionBlock.query.filter(SessionBlock.session_id.in_(sessions_ids))}
+
+    def get_validators(self, existing_registration):
+        def _check_number_of_sessions(new_data):
+            min_ = self.form_item.data.get('minimum')
+            max_ = self.form_item.data.get('maximum')
+            if max_ is not None and min_ is not None:
+                if max_ != 0 and len(new_data) > max_:
+                    raise ValidationError(_('Please select no more than {max} sessions.').format(max=max_))
+                if min_ != 0 and len(new_data) < min_:
+                    raise ValidationError(_('Please select at least {min} sessions.').format(min=min_))
+
+        def _check_session_block_is_valid(new_data):
+            old_selection = set()
+            if existing_registration and (old_data := existing_registration.data_by_field.get(self.form_item.id)):
+                old_selection = set(old_data.data)
+            event = self.form_item.registration_form.event
+            blocks = SessionBlock.query.filter(SessionBlock.id.in_(new_data)).all()
+            # disallow referencing invalid session blocks
+            if len(blocks) != len(new_data):
+                raise ValidationError('Cannot select session blocks which do not exist')
+            # disallow any session blocks from other events
+            if not all(sb.event == event for sb in blocks):
+                raise ValidationError('Cannot select session blocks from another event')
+            # ensure the user can access the session unless it was already selected before
+            if not all(sb.can_access(session.user) or sb.id in old_selection for sb in blocks):
```

Comment:
The second half of the condition is no longer needed, since we now always remove session block ids from the submitted value if they are not also included in the data received from the API endpoint.

I don't think this is a problem, since it's not logical for a manager to register someone for a session they cannot access anyway. If it is a problem (ping @jbtwist @omegak), we could pass the regdata to the API endpoint and make it return session details for restricted sessions as long as the user is currently registered for them. I would not add this extra complexity otherwise, because I think when someone is being registered for a restricted session they should also have access to it...

Reply:
The requirements from the UN do not include allowing managers to register participants for sessions that they cannot access. I can imagine cases where managers may want this but I would not add this complexity for the moment.